### DISCUSSION TITLE
ImPlot3D: An Extension of ImPlot for 3D Visualization

### DISCUSSION THREAD

**Discussion by brenocq** [epezent/implot#604] [Show and tell]
<p align="center">
<img src="https://github.com/user-attachments/assets/359473d2-73a9-452c-a5f3-cb96e3785dc2" width="270"> <img src="https://github.com/user-attachments/assets/97ec8be4-50f9-428b-b357-25e2479409b8" width="270"> <img src="https://github.com/user-attachments/assets/c212039b-4853-4d26-95a5-5470bf97555e" width="270">
</p>

<p align="center">
<img src="https://github.com/user-attachments/assets/ec7ec42a-3c62-44bf-9275-f735f0304c95" width="270"> <img src="https://github.com/user-attachments/assets/e6bd03fa-6d76-4f3e-8d15-c24a05a5f714" width="270"> <img src="https://github.com/user-attachments/assets/b66ff296-7fbf-4644-9129-37daecca0b62" width="270">
</p>

## 👋 Introducing [ImPlot3D](https://github.com/brenocq/implot3d)
Hey everyone!

I’ve been using ImPlot for years now, and I absolutely love it. It’s been an incredible tool for 2D data visualization, and it has helped me so much in my projects.

Over the past few months, my need to visualize 3D data started to grow, so I decided to implement [ImPlot3D](https://github.com/brenocq/implot3d). It’s heavily inspired by ImPlot (there’s a lot of copy-paste, haha) but extended to handle 3D visualizations. If you’re familiar with ImPlot, the API should feel very similar, just with an extra dimension!

ImPlot has been a huge inspiration for this project, and I wanted to share ImPlot3D with the community that has helped make ImPlot such an amazing tool. **It would be amazing to get some feedback**. Whether you’ve got ideas, suggestions, or just want to see what works (or breaks), I’d love to hear it!

## 🚀 Current Features
- Core 3D Plotting Functions:
  - `PlotScatter`: Plot individual 3D points.
  - `PlotLine`: Draw lines in 3D space.
  - `PlotTriangle`: Render triangles with fill support.
  - `PlotQuad`: Render quads (two triangles) for surfaces.
  - `PlotSurface`: Plot structured surfaces with X/Y grids.
  - `PlotMesh`: Display custom meshes using vertices and indices.
  - `PlotText`: Annotate your 3D plots with centered text labels.
- Interactivity:
  - Rotate, pan, and zoom plots using the mouse or touchpad.
  - Reset the camera view with double-click interactions.
  - Smooth camera rotation animations.
- Customization:
  - Control line styles, colors, and marker styles.
  - Apply colormaps to surfaces, lines, and markers.
  - Flexible axis setup with SetupAxes and SetupAxesLimits.

## 🛠️ Usage
The ImPlot3D API is designed to feel very similar to Dear ImGui and ImPlot. You start by calling `ImPlot3D::BeginPlot()` to initialize a 3D plot, followed by plotting various data using the `PlotX` functions (e.g., `PlotLine()` , `PlotScatter()` , `PlotSurface()` ). Finally, you end the plot with ` ImPlot3D::EndPlot()` .

```cpp
float x_data[1000] = ...;
float y_data[1000] = ...;
float z_data[1000] = ...;

ImGui::Begin("My Window");
if (ImPlot3D::BeginPlot("My Plot")) {
    ImPlot3D::PlotLine("My Line Plot", x_data, y_data, z_data, 1000);
    ImPlot3D::PlotScatter("My Scatter Plot", x_data, y_data, z_data, 1000);
    ...
    ImPlot3D::EndPlot();
}
ImGui::End();
```

## 🔗 Links
Repository: [ImPlot3D on GitHub](https://github.com/brenocq/implot3d)
Latest Release: [v0.1](https://github.com/brenocq/implot3d/releases/tag/v0.1)

ImPlot3D is still in its early stages. I’m excited to hear your thoughts and feedback as I continue to improve it! 🚀

**Comment by epezent** [comment]
@brenocq This is AMAZING! Fantastic work. I'm sure many ImPlot users will find this useful and it looks like you've done a great job staying consistent with ImGui and ImPlot's APIs. Can't wait to give it a go myself and look forward to following your progress. 

>  It’s heavily inspired by ImPlot (there’s a lot of copy-paste, haha)

On this, I'm curious if not depending on ImPlot directly was a deliberate decision or if there were limitations that made it impossible. For instance, if `implot.h` and/or `implot_internal.h` sufficiently exposed some of the inner workings, do you think there are areas the two libraries could converge (e.g. styling, types, utilities, shared implementation for widget behavior, etc.). Mostly just curious, and I recongize you answer might be due to ImPlot's lack of activity lately (my fault).  

PS -- I've linked ImPlot3D in our README under the FAQ. Cheers!

**Reply by brenocq** [reply]
Hey @epezent!

Thanks so much for adding ImPlot3D to the README so quickly—I really appreciate it!

Initially, I considered building ImPlot3D on top of ImPlot but decided to start from scratch since I wasn’t very familiar with the ImPlot codebase at the time. Along the way, I ended up reusing quite a bit of ImPlot's solutions for common problems I encountered, which was unexpected but made sense given the overlap in functionality. Now that I have a better understanding of ImPlot, I agree it would make a lot of sense to turn ImPlot3D into an extension of ImPlot to reduce duplicated work.

Many of ImPlot3D's internal structs already share fields with ImPlot, so refactoring ImPlot3D to rely more on ImPlot structs (with additions for the third axis) seems like a natural next step. Some adjustments on the ImPlot side would likely be needed too, so coordinating updates between the two repos will be important. It’s definitely something I’d love to collaborate on.

I also noticed the work on the backends branch—great stuff! It would be fantastic to make ImPlot3D faster. Right now, I’m using a custom draw list to sort triangles by depth before rendering, which is a bit costly. Using the depth buffer for handling triangle overlaps (at least for opaque surfaces) would be a much better solution and could eliminate the need for sorting altogether.

TL;DR: I think making ImPlot3D dependent on ImPlot is a great idea. It’ll take some refactoring on both sides, but it’s absolutely worth it to avoid duplicating work and improve maintainability. Let’s coordinate to make it happen! ❤️
